### PR TITLE
✨ Source Clockify: Add Optional API Url parameter

### PR DESCRIPTION
**Pull request by ezra-quemuel**
## What
Currently the source-clockify connector does not support subdomains for enterprise instances of clockify (e.g. mycompany.clockify.me)

#*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*
Add an optional parameter of api_url to source-clockify
See #27688 for more details
Screenshot 2023-06-24 at 7 00 19 PM

## How
*Describe the solution*
This solution adds an optional parameter of api_url with a default value of https://api.clockify.me and allows it to be customized.

## Recommended reading order
1.  source_clockify/spec.json
2. source_clockify/streams.py
3. source_clockify/source.py

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*
This PR contains no breaking changes. The version is minor since it adds an optional parameter of api url to the connector as per https://docs.airbyte.com/contributing-to-airbyte/#examples.

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by archangelic** [comment]
Thanks for your contribution! I'm going to review this and let you know if anything needs to be changed!

**Review by archangelic — CHANGES_REQUESTED** [review]
A few things that need to be fixed to continue:

The version for `dockerImageTag` in `metadata.yaml` needs to be updated.

In `setup.py` under `MAIN_REQUIREMENTS`, the airbyte-cdk requirement needs to be changed from `"airbyte-cdk~=0.2.0"` to `"airbyte-cdk"` so that it uses the latest version at all times.

**Comment by ezra-quemuel** [comment]
Thank you for the review, I have made the requested changes

File: airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 def setup_responses():
     responses.add(
         responses.GET,
-        "https://api.clockify.me/api/v1/workspaces/workspace_id/users",
```
Comment (sajarin) [inline airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py:14]: What's the reason for changing the unit test here?

**Review by sajarin — CHANGES_REQUESTED** [review]
Thanks for PR, do you have screenshots of passing integration tests?

**Comment by ezra-quemuel** [comment]
I've updated the schemas in response to api changes and have gotten integration tests to pass

Screenshot 2023-07-06 at 10 07 07 AM

File: airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py (hunk 1 above)
Comment (ezra-quemuel) [reply airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py:14]: I changed the base url for the tests in order to ensure that the Clockify API url is not hard coded

File: airbyte-integrations/connectors/source-clockify/source_clockify/streams.py [hunk 2]
```diff
@@ -13,13 +13,17 @@
...
     page_size = 50
     page = 1
     primary_key = None
 
-    def __init__(self, workspace_id: str, **kwargs):
+    def __init__(self, workspace_id: str, api_url: str, **kwargs):
         super().__init__(**kwargs)
+        self.api_url = api_url
+        self.url_base = self.api_url + self.api_path
         self.workspace_id = workspace_id
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/streams.py:28]: You can make this backward compatible if you transform the `url_base` to a property function in python.
Create a const to keep the previous base url.

        self.url_base = self.api_url + self.api_path
```suggestion
    def __init__(self, workspace_id: str, api_url: str, **kwargs):
        super().__init__(**kwargs)
        self.api_url = api_url or DEFAULT_API_URL
        self.workspace_id = workspace_id

    @property
    def url_base(self):
        return self.api_url + self.api_path
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-clockify/source_clockify/streams.py (hunk 2 above)
Comment (ezra-quemuel) [reply airbyte-integrations/connectors/source-clockify/source_clockify/streams.py]: Done!

**Comment by sajarin** [comment]
@ezra-quemuel thanks for the PR. Since this PR was created on your company's branch, I had to create a new branch and PR to run our CI and integration tests. 

These changes have been merged here: https://github.com/airbytehq/airbyte/pull/28622

**Pull request closed without merging**